In [175]:
#import pymongo as pm
import pandas as pd
import numpy as np
#import getpass
import os
import datetime
#from bson.objectid import ObjectId
import pprint
#from inspect import signature
import math
##import scipy.stats as stats
#from scipy.stats import normaltest
#from scipy.stats import kstest
#from distfit import distfit
#import array

#import seaborn as sns
import matplotlib.pyplot as plt
#import networkx as nx
#from cdlib import algorithms
#import random as rd
import re
from datetime import timedelta

In [157]:
import requests
payload = {
    "css_object_id": '002075129500006205772020', 
    "n_results": 10,
    "relevant_field": "description_problem_description", 
    "maximum_records": 10000, 
    "sbert_coeff": 200
}
fusion = requests.post('http://198.18.182.160/decision_fusion', json=payload)
fusion_json = fusion.json()

In [155]:
fusion_json

{'rank_of_rank_sum_result': [{'css_object_id': '002028376500002123832020',
   'customer_name': 'RHODIA OPERATIONS',
   'customer_no': '0000511168',
   'description_problem_description': 'SAP Hana Cockpit installation. Expert Area: SAP HANA Component: HAN-DB Dear * Support, We are planning to install SAP Hana Cockpit. Could you please provide details about couple of questions: SAP Hana Studio: when exactly sunset of this tool is planned (I just found info, that soon it will be substituted by SAP Hana Cockpit and that is all)? SAP HANA Cockpit Installation and Update Guide: Supported Hardware Platforms -> IBM Power Systems: SAP HANA cockpit is not supported for IBM Power systems running Red Hat Enterprise Linux. So, can we use IBM Power Systems to install SAP Hana Cockpit? and in such a case which version of OS needs to be used. In the newest version of Hana Cockpit in manifest file we have provided below information about revision "2.00.046". According to SAP Note 2235581 - SAP HANA: Su

In [86]:
from bs4 import BeautifulSoup
import json
def extract_customer_info(css_id):
    css_object_id = css_id
    print("Starting for " + css_object_id)
    r = requests.post('http://198.18.129.86/query_endpoint', data = {
        'query_id': css_object_id,
        'n_results': 10,
        'text_field': 'description_problem_description',
        'max_records': 10000,
        'sbert_coeff': 200,
        'backtrack_days': 60,
        'strong_threshold': 5,
        'weak_threshold': 3,
        'exclude_spam_solutions':False,
        'filter_by_prod_version':True,
        'filter_by_component':True
        })
    soup = BeautifulSoup(r.text)
    try:
        json_response = json.loads(soup.find(id="json_text").text)
        return json.dumps(json_response)
    except AttributeError:
        return ""

In [87]:
extract_customer_info('002075129500006205772020')

Starting for 002075129500006205772020


'{"Irrelevant": {"Customers": ["Colgate-Palmolive Company   Technol", "PepsiCo, Inc.", "Boston Scientific Corporation", "ADVANCED CHEMICAL INDUSTRIES LIMITE", "Salling Group", "Lemvigh-M\\u00fcller A/S", "Energy Transfer Operating, LP", "Consorcio Mexi-gas, S.A. de C.V.", "Zurich Insurance Company Ltd.", "The Coca-Cola Company", "AmorePacific Corporation", "SuccessFactors, Inc.", "Haworth, Inc.", "Bayer AG", "ServiceNow, Inc.", "Lufthansa AirPlus Servicekarten Gmb", "Trafikverket", "Woolworths Group Limited", "J&P Coats Ltd", "Ascend Performance Materials Holdin", "PAO Sberbank", "Alliander N.V.", "Cisco Systems, Inc.", "Otto-von-Guericke-Universit\\u00e4t Magde", "L Brands, Inc.", "GALP ENERGIA, S.A.", "CENTROS COMERCIALES CARREFOUR, S.A.", "Shoppers Stop Ltd", "NorgesGruppen ASA", "Anta (China) Co. Ltd.", "ADEO SERVICES SA", "KUBOTA Corporation", "Danfoss A/S", "Dimension Data Management Services", "Colt Technology Services", "Ceat Limited", "Contemporary Amperex Technology Co.", "NM

In [23]:
!pip3 install pymongo


Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/usr/bin/python3 -m pip install --upgrade pip' command.


In [74]:
import pymongo
import os
from pymongo import MongoClient
import pandas as pd
db_config= {"MONGODB_ADDRESS"    : os.environ.get('KEY_MONGO_ADDRESS','mongodb-service.stag-data.svc.cluster.local'),
            "MONGODB_PORT"       : os.environ.get('KEY_MONGO_PORT', '27017'),
            "MONGODB_USERNAME"   : os.environ.get('KEY_MONGO_USER', 'IBSO-read-write'),
            "MONGODB_PASSWORD"   : os.environ.get('KEY_MONGO_PASSWORD', 'NHWD5bw0xcMgy1Y'),
            "MONGODB_DATABASE"   : os.environ.get('KEY_MONGO_DATABASE', 'admin')}
mongo_client = MongoClient(                ('mongodb://%s:%s@' + db_config["MONGODB_ADDRESS"] + ':' + db_config["MONGODB_PORT"] + '/' + 
                                            db_config["MONGODB_DATABASE"]) % (                    db_config["MONGODB_USERNAME"], 
                                                                              db_config["MONGODB_PASSWORD"]), replicaset='rs0')
solutions_collection = mongo_client["main"]["solutions_parsed_version"]



SyntaxError: invalid character in identifier (<ipython-input-74-113d92cce0e0>, line 3)

In [58]:
from pymongo import MongoClient
import os as os
import pandas as pd
db_config= {
"MONGODB_ADDRESS"    : os.environ.get('KEY_MONGO_ADDRESS','mongodb-service.stag-data.svc.cluster.local'),
"MONGODB_PORT"       : os.environ.get('KEY_MONGO_PORT', '27017'),
"MONGODB_USERNAME"   : os.environ.get('KEY_MONGO_USER', 'IBSO-read-write'),
"MONGODB_PASSWORD"   : os.environ.get('KEY_MONGO_PASSWORD', 'NHWD5bw0xcMgy1Y'),
"MONGODB_DATABASE"   : os.environ.get('KEY_MONGO_DATABASE', 'admin')
}
mongo_client = MongoClient(
                ('mongodb://%s:%s@' + db_config["MONGODB_ADDRESS"] + ':' + db_config["MONGODB_PORT"] + '/' + db_config["MONGODB_DATABASE"]) % (
                    db_config["MONGODB_USERNAME"], db_config["MONGODB_PASSWORD"]), replicaset='rs0')
solutions_collection = mongo_client["main"]["solutions_parsed_version"]
solution_number = '2541754'
all_versions_list = list(solutions_collection.find({'Number': solution_number}).sort('ChangedOn_dt', -1))
all_versions_df = pd.DataFrame(all_versions_list)
latest_version = all_versions_list[0]



In [59]:
latest_version

{'_id': ObjectId('5f4e6203d1ae0500012d18de'),
 'AbapCode': [],
 'Attributes': [{'key': 'Responsible                                                                                         ',
   'Value': '8f41bb0beffaaf58d234743787e6cdc4459021bee0d02d7ab3b915610859d3a0'},
  {'key': 'Processor                                                                                           ',
   'Value': '8f41bb0beffaaf58d234743787e6cdc4459021bee0d02d7ab3b915610859d3a0'}],
 'Attributes_uh': [{'key': 'Responsible                                                                                         ',
   'Value': 'I306195'},
  {'key': 'Processor                                                                                           ',
   'Value': 'I306195'}],
 'Category': 'How To',
 'ChangedOn': '01.09.2020',
 'ChangedOn_dt': datetime.datetime(2020, 9, 1, 0, 0),
 'CorrectionInstructionsObjectList': [],
 'CreatedOn': '28.09.2017',
 'CreatedOn_dt': datetime.datetime(2017, 9, 28, 0, 0),
 'Current

In [60]:
import pandas as pd
df_all=pd.DataFrame(list(solutions_collection.find({}).limit(1)))

In [61]:
df_all.columns

Index(['_id', 'AbapCode', 'Attributes', 'Attributes_uh', 'Category',
       'ChangedOn', 'ChangedOn_dt', 'CorrectionInstructionsObjectList',
       'CreatedOn', 'CreatedOn_dt', 'CurrentStatus', 'ExpirationDate',
       'HasPreconditions', 'InternalMemo', 'Languages', 'Number',
       'OtherSAPComponents', 'Patches', 'PrimarySAPComponent', 'Priority',
       'Product', 'RatingInfo', 'RefNumber', 'ReferenceTo', 'ReferencedBy',
       'ReleasedOn', 'ReleasedOn_dt', 'SoftwareComponents', 'SupportPackages',
       'TargetStatus', 'TextSections', 'Title', 'TranslationFlag', 'Type',
       'Version', 'extracted_at', 'extracted_at_dt', 'sha_hashed', 'Symptom',
       'Solution', 'Reason&Prerequisites', 'KeyTerms', 'ReproducingTheIssue',
       'Cause', 'AbapCode_parsed', 'NormalizedTitle', 'NormalizedSymptom',
       'NormalizedSolution', 'NormalizedReason&Prerequisites',
       'NormalizedKeyTerms', 'NormalizedReproducingTheIssue',
       'NormalizedCause', 'NormalizedText', 'TermCount', 'par

In [52]:
solution_data = pd.DataFrame(list(solutions_collection.find({},{'_id':0,'Number':1,'Version':1, 'Category':1, 'ChangedOn_dt':1, 
                                                                'CreatedOn_dt':1, 'PrimarySAPComponent':1, 'OtherSAPComponents':1,
                                                                'ReleasedOn_dt':1,  'Priority':1,
                                                                'Product':1, 'Type':1,  'Title':1,
                                                                'Symptom':1, 'Solution':1, 'KeyTerms':1,'TermCount':1
                                                               })))

In [55]:
solution_data.shape

(2676523, 16)

The Solution Data has 2,676,523 pages - some pages hae multiple versions

In [54]:
solution_data.head(10)

Category ChangedOn_dt         CreatedOn_dt Number OtherSAPComponents  \
0            1992-11-11                        1180                 []   
1            1992-11-12                         762                 []   
2            1992-11-13                         768                 []   
3            1992-12-16  1992-12-10 00:00:00   2908                 []   
4            1992-04-10                          44                 []   
5            1992-08-10  1992-08-10 00:00:00   1862                 []   
6            1992-11-12                         733                 []   
7            1992-10-20  1992-10-20 00:00:00   2413                 []   
8            1992-03-02                         126                 []   
9            1992-10-23  1992-10-23 00:00:00   2441                 []   

  PrimarySAPComponent Priority Product ReleasedOn_dt  \
0              bc-aba               []    1992-11-13   
1              bc-aba               []    1992-11-12   
2              bc-aba               []    1992-11-13   
3           bc-db-dbi               []    1992-12-16   
4                  bc               []    1992-04-10   
5              bc-aba               []    1992-11-13   
6              bc-aba               []    1992-11-12   
7                  bc               []    1992-10-20   
8              bc-aba               []    1992-03-02   
9                  bc               []    1992-10-23   

                                              Title  Type Version  \
0  Unexpected field contents w/IMPORT .. /SELECT ..  note    0001   
1   Offset and length of function module parameters  note    0001   
2      Unfulfillable sort conditions for SAP-SELECT  note    0001   
3  Time difference in TWO_TASK connection to Oracle  note    0001   
4                CUA interface: multiple user logon  note    0001   
5       Necessary line width for GI or array-DELETE  note    0001   
6         Error message INSERT_REPORT_LINE_TOO_LONG  note    0001   
7               Problems when ENTER key is assigned  note    0001   
8    Sequence of records with  SELECT ... ENDSELECT  note    0001   
9                  Error with parameter transaction  note    0001   

                                             Symptom  \
0  Key word: ABAP/4 When filling an internal tabl...   
1  Key word: PARAMETER The syntax check reports t...   
2  Key word: ABAP/4 A SAP-SELECT statement in the...   
3  Key word: Time diff. A time difference has bee...   
4  Key word: CUA logon In the 3270 environment, m...   
5  Key word: ABAP/4 With GI-DELETE "DELETE dbtab"...   
6  Key word: Short dump Abend: report has line wi...   
7  Key word: Batch input The transaction does not...   
8  Key word: ABAP/4 A SELECT instruction without ...   
9  Key word: Parameter The parameters defined in ...   

                                            Solution KeyTerms TermCount  
0  After an IMPORT or SELECT in a target table wi...                124  
1  Alter programs such that no offset or length s...                 37  
2  In order to perform a sort by the primary key,...                108  
3  When installing the SAP System, the same time ...                 50  
4  Is designed this way. a user can log on to the...                 50  
5  Suggestion for enhancement recorded in develop...                 48  
6         Only fill tables line up to 72 characters.                 28  
7  The batch input programs should transmit the p...                120  
8  Add an ORDER-BY clause to the SELECT instruction.                 32  
9  The transactions triggered via TSTCP must not ...                 55

In [22]:
solution_data.sort_values(by=['Number']).head(100)

Category ChangedOn_dt         CreatedOn_dt  \
1263447                 Program error   2013-11-09  1998-03-31 00:00:00   
1722909                    Consulting   2017-01-16  2006-11-16 00:00:00   
1301876                 Program error   2013-11-12  2006-11-16 00:00:00   
1301901  Correction of legal function   2013-11-12  2006-11-16 00:00:00   
710970                     Consulting   2008-05-29  2006-11-16 00:00:00   
...                               ...          ...                  ...   
552001                  Program error   2006-11-16  2006-11-16 00:00:00   
572266                  Program error   2007-02-19  2006-11-16 00:00:00   
552530                  Program error   2006-11-17  2006-11-16 00:00:00   
1302021                 Program error   2013-11-12  2006-11-16 00:00:00   
991320                     Consulting   2011-05-13  2006-11-16 00:00:00   

          Number PrimarySAPComponent                           Priority  \
1263447   100000            sd-bf-ac    Correction with medium priority   
1722909  1000000           bc-wd-aba  Recommendations / Additional Info   
1301876  1000002    is-a-dbm-ord-wty    Correction with medium priority   
1301901  1000003           xx-csc-ar    Correction with medium priority   
710970   1000004  bw-bex-ot-olap-aut  Recommendations / Additional Info   
...          ...                 ...                                ...   
552001   1000094         is-r-ifc-so      Correction with high priority   
572266   1000095           crm-md-li      Correction with high priority   
552530   1000096          fi-gl-gl-d       Correction with low priority   
1302021  1000097     bw-whm-dba-iobj    Correction with medium priority   
991320   1000098     ppm-pro-ext-pmt  Recommendations / Additional Info   

        Product ReleasedOn_dt  \
1263447      []    1999-08-19   
1722909      []    2017-01-12   
1301876      []    2006-11-20   
1301901      []    2006-11-17   
710970       []    2008-05-29   
...         ...           ...   
552001       []    2006-11-16   
572266       []    2007-02-19   
552530       []    2006-11-16   
1302021      []    2006-11-16   
991320       []    2006-11-16   

                                                     Title  Type Version  \
1263447                         Contracts and requirements  note    0004   
1722909                                Web Dynpro ABAP FAQ  note    0014   
1301876        Duration of doing a recall is not displayed  note    0002   
1301901  J_1AF217: Over double click wrong document is ...  note    0002   
710970      Merging and optimizing analysis authorizations  note    0006   
...                                                    ...   ...     ...   
552001   Transferring quantities and units of measure o...  note    0002   
572266   Exception in Listings on Back navigation from ...  note    0002   
552530   FAGLB03: Message F5 165 after selecting 'Indiv...  note    0001   
1302021  BI Content reconcilliation for InfoObjs termin...  note    0001   
991320   Export/import import function and status-depen...  note    0004   

                                                   Symptom  \
1263447  If the settings in Customizing for contracts a...   
1722909  This note provides answers to frequently asked...   
1301876       Duration of doing a recall is not displayed.   
1301901  Wrong document is fetched when a double click ...   
710970   This note describes how to merge and optimize ...   
...                                                    ...   
552001   When you create a purchase order using a store...   
572266   In the Listings Application when the user navi...   
552530   You use transaction FAGLB03 to display the bal...   
1302021  You compare and reconcile an InfoObject with y...   
991320   You release a task and you manually maintain f...   

                                                  Solution  \
1263447                           Correct the source code.   
1722909  Question: When I click to create a roun

In [30]:
n = len(pd.unique(solution_data['Version']))
  
print("No.of.unique values :", 
      n)

No.of.unique values : 1014


The maximum number of versions a Solution page has is 1014 

In [29]:
pd.unique(solution_data['Type'])

array(['note', 'kba', 'b1_note', 'SAP Security Note'], dtype=object)

In [57]:
solution_data['Type'].value_counts()

note                 2292897
kba                   362762
b1_note                20302
SAP Security Note        562
Name: Type, dtype: int64

In [163]:
pd.unique(solution_data['Category'])

array(['', 'Program error', 'Upgrade information', 'Advance development',
       'Consulting', 'Customizing', 'Workaround of missing functionality',
       'Help for error analysis', 'Special development', 'Performance',
       'Documentation error', 'Installation information',
       'Release planning information', 'Translation error',
       'External error', 'How To', 'Legal change', 'Problem', 'Bug Filed',
       'Exit added', 'Product Enhancement', 'FAQ',
       'Correction of legal function', 'Year 2000',
       'Announcement of Legal Change', 'Modification', 'Known Error'],
      dtype=object)

In [56]:
solution_data['Category'].value_counts()

Program error                          1843166
Problem                                 207038
How To                                  123467
Correction of legal function             78636
Consulting                               68241
Advance development                      45358
Performance                              38287
Legal change                             35242
Installation information                 27492
Customizing                              26018
Upgrade information                      21156
Bug Filed                                19519
FAQ                                      18900
Special development                      18193
Help for error analysis                  17171
Release planning information             17011
Translation error                        13460
Documentation error                      12065
Workaround of missing functionality      10921
Product Enhancement                      10620
Exit added                                7418
Modification 

In [165]:
solution_data.hist(column = 'Category')

ValueError: hist method requires numerical columns, nothing to plot.

In [52]:
pd.unique(solution_data['PrimarySAPComponent'])

array(['bc-aba', 'bc-db-dbi', 'bc', ..., 'lod-sf-rnr-mil',
       'lod-sf-cpm-chk', 'scm-wr'], dtype=object)

In [53]:
len(pd.unique(solution_data['PrimarySAPComponent']))

17312

In [31]:
pd.unique(solution_data['Version'])

array(['0001', '0002', '0003', ..., '0877', '0994', '1033'], dtype=object)

In [50]:
len(pd.unique(solution_data['Number']))

2395747

The number of unique solutions is 2,395,747 ???

In [ ]:
pd.unique(solution_data['Number'])[1:10]

In [32]:
solution_data[solution_data['Version'] == '1033']

Category ChangedOn_dt         CreatedOn_dt   Number  \
2673507   How To   2022-05-15  2014-04-02 00:00:00  1999997   

        PrimarySAPComponent Priority  \
2673507              han-db   Normal   

                                            Product ReleasedOn_dt  \
2673507  [SAP HANA, platform edition all versions ]    2022-05-15   

                        Title Type Version  \
2673507  FAQ: SAP HANA Memory  kba    1033   

                                                   Symptom  \
2673507  You have questions related to the SAP HANA mem...   

                                                  Solution  \
2673507  1. Which indications exist for SAP HANA memory...   

                                                  KeyTerms TermCount  
2673507  SAP HANA memory heap allocator table row colum...     22266

In [33]:
solution_data[solution_data['Number'] == '1999997']

Category ChangedOn_dt         CreatedOn_dt   Number  \
2117600   How To   2019-07-30  2014-04-02 00:00:00  1999997   
2145816   How To   2019-09-07  2014-04-02 00:00:00  1999997   
2146753   How To   2019-09-10  2014-04-02 00:00:00  1999997   
2147207   How To   2019-09-10  2014-04-02 00:00:00  1999997   
2150704   How To   2019-09-14  2014-04-02 00:00:00  1999997   
...          ...          ...                  ...      ...   
2622702   How To   2022-01-03  2014-04-02 00:00:00  1999997   
2626595   How To   2022-01-12  2014-04-02 00:00:00  1999997   
2639465   How To   2022-04-01  2014-04-02 00:00:00  1999997   
2655688   How To   2022-04-10  2014-04-02 00:00:00  1999997   
2673507   How To   2022-05-15  2014-04-02 00:00:00  1999997   

        PrimarySAPComponent Priority  \
2117600              han-db   Normal   
2145816              han-db   Normal   
2146753              han-db   Normal   
2147207              han-db   Normal   
2150704              han-db   Normal   
...                     ...      ...   
2622702              han-db   Normal   
2626595              han-db   Normal   
2639465              han-db   Normal   
2655688              han-db   Normal   
2673507              han-db   Normal   

                                            Product ReleasedOn_dt  \
2117600  [SAP HANA, platform edition all versions ]    2019-07-30   
2145816  [SAP HANA, platform edition all versions ]    2019-09-07   
2146753  [SAP HANA, platform edition all versions ]    2019-09-10   
2147207  [SAP HANA, platform edition all versions ]    2019-09-10   
2150704  [SAP HANA, platform edition all versions ]    2019-09-14   
...                                             ...           ...   
2622702  [SAP HANA, platform edition all versions ]    2022-01-03   
2626595  [SAP HANA, platform edition all versions ]    2022-01-12   
2639465  [SAP HANA, platform edition all versions ]    2022-04-01   
2655688  [SAP HANA, platform edition all versions ]    2022-04-10   
2673507  [SAP HANA, platform edition all versions ]    2022-05-15   

                        Title Type Version  \
2117600  FAQ: SAP HANA Memory  kba    0778   
2145816  FAQ: SAP HANA Memory  kba    0779   
2146753  FAQ: SAP HANA Memory  kba    0780   
2147207  FAQ: SAP HANA Memory  kba    0781   
2150704  FAQ: SAP HANA Memory  kba    0782   
...                       ...  ...     ...   
2622702  FAQ: SAP HANA Memory  kba    1020   
2626595  FAQ: SAP HANA Memory  kba    1021   
2639465  FAQ: SAP HANA Memory  kba    1029   
2655688  FAQ: SAP HANA Memory  kba    1030   
2673507  FAQ: SAP HANA Memory  kba    1033   

                                                   Symptom  \
2117600  You have questions related to the SAP HANA mem...   
2145816  You have questions related to the SAP HANA mem...   
2146753  You have questions related to the SAP HANA mem...   
2147207  You have questions related to the SAP HANA mem...   
2150704  You have questions related to the SAP HANA mem...   
...                                                    ...   
2622702  You have questions related to the SAP HANA mem...   
2626595  You have questions related to the SAP HANA mem...   
2639465  You have questions related to the SAP HANA mem...   
2655688  You have questions related to the SAP HANA mem...   
2673507  You have questions related to the SAP HANA mem...   

                                                  Solution  \
2117600  1. Which indications exist for SAP HANA memory...   
2145816  1. Which indications exist for SAP HANA memory...   
2146753  1. Which indications exist for SAP HANA memory...   
2147207  1. Which indications exist for SAP HANA memory...   
2150704  1. Which indications exist for SAP HANA memory...   
...                                                    ...   
2622702  1. Which indications exist for SAP HANA memory...   
2626595  1. Which indications exist for SAP HANA memory...   
2639465  1. Which indications exist for SAP HANA memory...   
2655688  1. Which indic

In [37]:
solution_data['Version'] = solution_data['Version'].astype(int)


In [42]:
#solution_data_grouped - []
#solution_data_grouped = solution_data.groupby('Number').max('Version')

idx = solution_data.groupby(['Number'])['Version'].transform(max) == solution_data['Version']

solution_data_grouped = solution_data[idx]

In [16]:
df_solution_data_grouped = pd.DataFrame(solution_data_grouped)

In [30]:
type(solution_data_grouped)

pandas.core.frame.DataFrame

In [28]:
solution_data_grouped.shape
list(solution_data_grouped.columns)

['Category',
 'ChangedOn_dt',
 'CreatedOn_dt',
 'Number',
 'OtherSAPComponents',
 'PrimarySAPComponent',
 'Priority',
 'Product',
 'ReleasedOn_dt',
 'Title',
 'Type',
 'Version',
 'Symptom',
 'Solution',
 'KeyTerms',
 'TermCount']

In [33]:
solution_data_grouped['Number'][1001:1020]

1001    15451
1002    25661
1003    27345
1004    31212
1005    25856
1006    30775
1007    31004
1008    30955
1009    30909
1010    30349
1011    30372
1012    30448
1013    29647
1014    26835
1015    27603
1016    28955
1017    30766
1018    31053
1019    30649
Name: Number, dtype: object

In [31]:
solution_data_grouped["PrimarySAPComponent"][1001:1020]

1001     fi-aa-aa-f
1002    co-om-cca-b
1003       lo-md-mm
1004          qm-im
1005     fi-aa-aa-b
1006             fi
1007             fi
1008             qm
1009     fi-sl-is-a
1010             sd
1011             fi
1012             sd
1013       mm-im-pi
1014             co
1015    co-om-cca-b
1016             fi
1017       fi-aa-is
1018         sd-sls
1019             mm
Name: PrimarySAPComponent, dtype: object

In [32]:
solution_data_grouped["OtherSAPComponents"][1001:1020]

1001                                       []
1002                                       []
1003                                       []
1004                                       []
1005                                       []
1006                                       []
1007                                       []
1008                                  [QM-IM]
1009                                       []
1010                                       []
1011                                       []
1012                                       []
1013                                       []
1014    [CO-OM-CCA-G, CO-OM-OPA-G, PS-IS-ACC]
1015                                       []
1016                                       []
1017                                       []
1018                              [SD-SLS-SO]
1019                               [LO-MD-MM]
Name: OtherSAPComponents, dtype: object

In [62]:





metadata_collection = mongo_client["IBSO"]["infodocs_metadata"]
css_object_id = '002075129400003820012020'
metadata = metadata_collection.find_one({'css_object_id': css_object_id})


In [63]:
metadata

{'_id': ObjectId('611d999a93044c203acf1ed2'),
 'component': 'SCM-EWM-CNT',
 'css_object_id': '002075129400003820012020',
 'customer_name': 'Voss Automotive GmbH',
 'customer_no': '0000028770',
 'description': 'Requesting further Q system provisioning & activation',
 'description_problem_description': 'Requesting further Q system provisioning & activation. --- Production --- Product selected: SAP S/4HANA Cloud Component selected (Manual Selection): XX-S4C-OPR-SRV --- Steps to Reproduce --- Note to SAP SUPPORT: I consent to SAP Support reproducing the described issue while connected remotely to my non-production environments, using any steps provided by me in the description below. If it is necessary to do this in a production environment, SAP Support will contact me specifically before proceeding. --- Description --- Hello Experts, Please activate the Q-system. Attached are the required Q-system Activation Request Template. Below are the additional comments for activation. Also addition

In [101]:
df_solution_data_grouped['KeyTerms']

0                                                           
1                                                           
2                                                           
3                                                           
4                                                           
                                 ...                        
2676518                                                     
2676519    SHA-256,Outbound SSO,Solex,BenefitFocus,Skills...
2676520                                                     
2676521    SAP Digital Manufacturing Cloud, SAP Digital M...
2676522                                                     
Name: KeyTerms, Length: 2441955, dtype: object

In [176]:
from pymongo import MongoClient
db_config= {
"MONGODB_ADDRESS"    : os.environ.get('KEY_MONGO_ADDRESS','mongodb-service.stag-data.svc.cluster.local'),
"MONGODB_PORT"       : os.environ.get('KEY_MONGO_PORT', '27017'),
"MONGODB_USERNAME"   : os.environ.get('KEY_MONGO_USER', 'IBSO-read-write'),
"MONGODB_PASSWORD"   : os.environ.get('KEY_MONGO_PASSWORD', 'NHWD5bw0xcMgy1Y'),
"MONGODB_DATABASE"   : os.environ.get('KEY_MONGO_DATABASE', 'admin')
}
mongo_client = MongoClient(
                ('mongodb://%s:%s@' + db_config["MONGODB_ADDRESS"] + ':' + db_config["MONGODB_PORT"] + '/' + db_config["MONGODB_DATABASE"]) % (
                    db_config["MONGODB_USERNAME"], db_config["MONGODB_PASSWORD"]), replicaset='rs0')
infodocs_metadata = mongo_client["IBSO"]["infodocs_metadata"]
metadata_collection = mongo_client["IBSO"]["infodocs_metadata"]

In [37]:
df_infodocs_metadata = pd.DataFrame(infodocs_metadata.find({"incident_created_at_dt":{
   "$gte": datetime.datetime(2022,2,1),
    "$lt": datetime.datetime(2022,4,1)
}}))

In [38]:
df_infodocs_metadata.head(10)

_id             css_object_id incident_created_at_dt  \
0  62681b684cc370bff2568330  002024211700000044652022    2022-02-01 00:01:19   
1  62681b684cc370bff25696ef  002075129400000737852022    2022-02-01 00:04:31   
2  62681b684cc370bff25696e0  002075129400000737372022    2022-02-01 00:04:31   
3  62681b684cc370bff2568331  002024211700000044662022    2022-02-01 00:06:16   
4  62681b684cc370bff2568337  002024211700000044772022    2022-02-01 00:06:21   
5  62681b684cc370bff256833b  002024211700000044852022    2022-02-01 00:08:58   
6  62681b684cc370bff25696f1  002075129400000737962022    2022-02-01 00:11:29   
7  62681b684cc370bff256b04f  002075129500000820532022    2022-02-01 00:14:24   
8  62681b684cc370bff25696f0  002075129400000737862022    2022-02-01 00:15:29   
9  62681b684cc370bff25696f4  002075129400000738012022    2022-02-01 00:15:29   

                         customer_name customer_no        component  \
0         VISTRA HOLDINGS (UK) LIMITED  0002054244   XX-S4C-OPR-SRV   
1  Distribuidora de Electricidad del S  0000875428      MOB-APP-MAO   
2                       Jemena Limited  0000237010    LOD-SF-EC-FOO   
3      Mundipharma IT Services Limited  0001572344       XX-HST-OPR   
4                     Mann+Hummel GmbH  0000011597        FI-CNT-GL   
5                            SAP Cloud  0001119571  SCM-IBP-OPS-SIZ   
6  Luminous Power Technologies Pvt Ltd  0001055712   XX-HST-OPR-INC   
7           Southern California Edison  0000718736      HAN-DB-PERF   
8       Department of Customer Service  0000691023        FI-TV-COS   
9  Empresa Colombiana De Petroleos S.A  0000671855   LOD-SF-LMS-INT   

                                         description  \
0          Foreign Key Errors Present in Q2P Project   
1  Error transmitting meter movement from Work Ma...   
2  SM - Unable to maintain translation for Founda...   
3                            PS4MUN - System is down   
4          Foreign Key Errors Present in Q2P Project   
5                                OOM events per week   
6  Getting error code 9603 , communication link f...   
7  SAP GRC EAM on HANA DB - System slowness due t...   
8               Credit Card receipts getting doubled   
9                    Administrators LMS access error   

                     prod_version  \
0          SAP S/4HANA CLOUD 2111   
1         SAP MOBILE PLATFORM 3.0   
2   SUCCESSFACTORS BIZX CORE 2111   
3                SAP S/4HANA 1909   
4          SAP S/4HANA CLOUD 2111   
5                 SAP IBP OD 2111   
6                    SAP TDMS 4.0   
7                 NW AS ABAP 7.52   
8                SAP S/4HANA 2020   
9  SUCCESSFACTORS ONBOARDING 2111   

                                  communication_logs  \
0  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
1  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
2  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
3  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
4  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
5  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
6  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
7  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
8  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   
9  [{'datetime': '', 'text': '<p>', 'type': ''}, ...   

                                               solns  \
0                                            [10, 5]   
1                                       [0002504590]   
2                           [0002197595, 0002079466]   
3                                       [0000000010]   
4                                               [10]   
5                                               [10]   
6                                       [0000000010]   
7  [0002000000, 0001969700, 0001999993, 000218016...   
8                           [0001239839, 0001523821]   
9                                       [0002264167]   

                     description_problem_description  \
0  Foreign Key Errors Present in Q2P Project. 

In [154]:
def preprocess_txt(txt):
    txt = txt.lower()
    #punctuations = ''','''
    punctuations = '''!()[]{};:'"\,<>./?@#$%^&*_~'''
    for x in txt:
        if x in punctuations:
            txt = txt.replace(x, " ")
    txt = txt.split(" ")
    txt = ' '.join(x for x in txt if len(x)>=3)
    #txt = re.sub(r'\d',"#", txt)
    return(txt)
        

In [138]:
df_solution_data_grouped['KeyTerms'][1011:1012]

1011    FS10 SAPMF42B T001A GD13
Name: KeyTerms, dtype: object

In [155]:
solutionKeyTermsText = []
for d in df_solution_data_grouped['KeyTerms'][1001:1020]:
    #print(d)
    content = preprocess_txt(d)
    print(content)
    solutionKeyTermsText.append(content)



program sapmgplv transactions kp96 kp97
transactions mm01 mm02 program sapmm03m
qa32 qa33
rabest01 ragitt01 razuga01 raabga01 rahafa01 sapdbada

fb01
oqv1 tq07 tq07a
report writer

fs10 sapmf42b t001a gd13
condition tables

transactions ksb1 kob1 ksb2 kob2 cj74 cj76
program sapmgplv background processing
sapf110s group currency payment proposal payment run payment document
ragitt01 intangible complex fixed asset low-value asset ica lva




In [156]:
solutionKeyTermsText

['',
 'program sapmgplv transactions kp96 kp97',
 'transactions mm01 mm02 program sapmm03m',
 'qa32 qa33',
 'rabest01 ragitt01 razuga01 raabga01 rahafa01 sapdbada',
 '',
 'fb01',
 'oqv1 tq07 tq07a',
 'report writer',
 '',
 'fs10 sapmf42b t001a gd13',
 'condition tables',
 '',
 'transactions ksb1 kob1 ksb2 kob2 cj74 cj76',
 'program sapmgplv background processing',
 'sapf110s group currency payment proposal payment run payment document',
 'ragitt01 intangible complex fixed asset low-value asset ica lva',
 '',
 '']

In [141]:

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

ModuleNotFoundError: No module named 'sklearn'

In [143]:
%pip install sklearn --user

  Preparing metadata (setup.py) ... done
     |████████████████████████████████| 22.2 MB 17.3 MB/s            
     |████████████████████████████████| 306 kB 55.8 MB/s            
  Created wheel for sklearn: filename=sklearn-0.0-py2.py3-none-any.whl size=2397 sha256=b791ee0cb4573a16a6eea154a2bcd6bcc52602c562ca5d048a135599d485b091
  Stored in directory: /home/jovyan/.cache/pip/wheels/23/9d/42/5ec745cbbb17517000a53cecc49d6a865450d1f5cb16dc8a9c
Successfully built sklearn
Note: you may need to restart the kernel to use updated packages.


In [144]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [148]:
solutions_vectorizer = TfidfVectorizer(solutionKeyTermsText)
solutiona_KeyTerms_tfidf = solutions_vectorizer.fit_transform(solutionKeyTermsText)

/home/jovyan/.local/lib/python3.6/site-packages/sklearn/utils/validation.py:72: FutureWarning: Pass input=['', 'program  sapmgplv transactions  kp96  kp97', 'transactions  mm01  mm02  program  sapmm03m', 'qa32  qa33', 'rabest01 ragitt01 razuga01 raabga01 rahafa01 sapdbada', '', 'fb01', 'oqv1  tq07  tq07a', 'report writer', '', 'fs10 sapmf42b t001a gd13', 'condition tables', '', 'transactions  ksb1  kob1  ksb2  kob2  cj74  cj76', 'program  sapmgplv background processing', 'sapf110s  group currency  payment proposal  payment run  payment document', 'ragitt01 intangible complex fixed asset  low-value asset ica lva', '', ''] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error
  "will result in an error", FutureWarning)


In [146]:
solutions_vectorizer


TfidfVectorizer(input=['', 'program  sapmgplv transactions  kp96  kp97',
                       'transactions  mm01  mm02  program  sapmm03m',
                       'qa32  qa33',
                       'rabest01 ragitt01 razuga01 raabga01 rahafa01 sapdbada',
                       '', 'fb01', 'oqv1  tq07  tq07a', 'report writer', '',
                       'fs10 sapmf42b t001a gd13', 'condition tables', '',
                       'transactions  ksb1  kob1  ksb2  kob2  cj74  cj76',
                       'program  sapmgplv background processing',
                       'sapf110s  group currency  payment proposal  payment '
                       'run  payment document',
                       'ragitt01 intangible complex fixed asset  low-value '
                       'asset ica lva',
                       '', ''])

In [149]:
solutiona_KeyTerms_tfidf

<19x51 sparse matrix of type '<class 'numpy.float64'>'
	with 57 stored elements in Compressed Sparse Row format>

In [157]:
def get_tf_idf_query_similarity(vectorizer, docs_tfidf, query):
    """
    vectorizer: TfIdfVectorizer model
    docs_tfidf: tfidf vectors for all docs
    query: query doc

    return: cosine similarity between query and all docs
    """
    query_tfidf = vectorizer.transform([query])
    cosineSimilarities = cosine_similarity(query_tfidf, docs_tfidf).flatten()
    return cosineSimilarities


In [158]:
get_tf_idf_query_similarity(solutions_vectorizer, solutiona_KeyTerms_tfidf, 'program  sapmgplv transactions  kp96  kp97')

array([0.        , 1.        , 0.30219223, 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.1210056 , 0.37744273,
       0.        , 0.        , 0.        , 0.        ])

# Try the ISM API - first using the FAST API implementation

In [158]:
import requests
payload = {
    "queryId": "002028376500002067582021",
    "userId": "D123456",
    "systemId": "A11",
    "sessionId": "961779bd-726b-b62c-d035-73906fbf35db",
    "customerNumber": "S000000001",
    "customerName": "SAP AG",
    "installationNumber": "0020982966",
    "leadingProduct": " SAP NetWeaver  > Application Server ABAP ",
    "swType": "ONPREM",
    "swClass": "",
    "title": "SAP NetWeaver Application Server for ABAP",
    "problemDescription": "SAP NetWeaver Application Server for ABAP",
    "stepsToReproduce": "do this, then that",
    "refinementText": "more information here",
    "component": "BW-BEX-ET",
    "firstEntry": 0,
    "lastEntry": 40,
    "systemRole": "123",
    "otherCommunication": [
    {
      "date": "20210304T123001",
      "type": "refToSAP",
      "text": "setup wizard"
    },
    {
      "date": "20210304T123105",
      "type": "internalMemo",
      "text": "sample memo"
    }
    ],
    "supportPackages": [
    "SAPK-60853INSAPHRCSI",
    "SAPK-60853INSAPHRCSK",
    "SAPK-60853INSAPHRCTR"
    ],
    "productVersion": "SAP NetWeaver 7.2",
    "FilterSolutionType": [
    "note",
    "kba"
    ],
    "FilterSolutionStatus": [
    "Released for Customer",
    "Released Internally"
    ],
    "suggestKeywords": False
}
res = requests.post('http://ism-api-service-dev.ism-api.svc.cluster.local/',json=payload)
res_json = res.json()
res_json

{'result_list': [{'nr': '3188945',
   'score': 74.41167865010874,
   'type': 'Note',
   'status': 'Released for Customer',
   'summary': 'Problems with the UI theme designer occur in SAP NetWeaver Application Server ABAP (SAP NetWeaver AS ABAP).',
   'component': 'CA-UI2-THD',
   'title': 'UI Theme Designer on ABAP: Version 1.70.0',
   'released_at': '2022-04-21T00:00:00',
   'trending': False,
   'category': 'Program error',
   'age': 'Released 7 months ago',
   'url': 'https://api.ait.c.eu-de-2.cloud.sap/ism-now-redirector/v0/note/?nr=3188945&incidentID=002028376500002067582021&refine=more information here&ISM_rec=3188945,3191869,3126234,3082432,3189594,3016569,3007764,3027560,2994291,3007076',
   'external_url': 'https://launchpad.support.sap.com/#/notes/3188945',
   'componentLong': ''},
  {'nr': '3191869',
   'score': 74.40161828815187,
   'type': 'KBA',
   'status': 'Released to Customer',
   'summary': 'You are curious whether your SAP NetWeaver Application Server ABAP system is

In [159]:
df_res = pd.json_normalize(res_json, record_path = ['result_list'])

In [160]:
df_res.shape

(40, 14)

In [150]:
customerID = "002028376500002067582021"
solutionID = "3188945"
df_res.loc[df_res['nr'] == solutionID].index[0]

0

In [151]:
def get_solution_rank(customerID, solutionID):
    """
    return: rank of solutionID in the list of solutions proposed by the ISM API for customerID
    """
    
    payload = {
    "queryId": customerID,
    "userId": "D123456",
    "systemId": "A11",
    "sessionId": "961779bd-726b-b62c-d035-73906fbf35db",
    "customerNumber": "S000000001",
    "customerName": "SAP AG",
    "installationNumber": "0020982966",
    "leadingProduct": " SAP NetWeaver  > Application Server ABAP ",
    "swType": "ONPREM",
    "swClass": "",
    "title": "SAP NetWeaver Application Server for ABAP",
    "problemDescription": "SAP NetWeaver Application Server for ABAP",
    "stepsToReproduce": "do this, then that",
    "refinementText": "more information here",
    "component": "BW-BEX-ET",
    "firstEntry": 0,
    "lastEntry": 10,
    "systemRole": "123",
    "otherCommunication": [
    {
      "date": "20210304T123001",
      "type": "refToSAP",
      "text": "setup wizard"
    },
    {
      "date": "20210304T123105",
      "type": "internalMemo",
      "text": "sample memo"
    }
    ],
    "supportPackages": [
    "SAPK-60853INSAPHRCSI",
    "SAPK-60853INSAPHRCSK",
    "SAPK-60853INSAPHRCTR"
    ],
    "productVersion": "SAP NetWeaver 7.2",
    "FilterSolutionType": [
    "note",
    "kba"
    ],
    "FilterSolutionStatus": [
    "Released for Customer",
    "Released Internally"
    ],
    "suggestKeywords": False
    }
    res = requests.post('http://ism-api-service-dev.ism-api.svc.cluster.local/',json=payload)
    res_json = res.json()
    df_res = pd.json_normalize(res_json, record_path = ['result_list'])
    rank_index = df_res.loc[df_res['nr'] == solutionID].index[0] + 1
    return rank_index

In [152]:
get_solution_rank(customerID, solutionID)

1

# Get CSS Incident ID for a doublet

In [177]:
# To get latest css_object_id for a doublet
customer_no = '0002348832'
installation_no = '0090463694'
query = {
    'customer_no': customer_no,
    'installation_no': installation_no
}
res_list = list(metadata_collection.find(query).sort('incident_created_at_dt', -1).limit(1))
if len(res_list) == 0:
    print('No doublet match found')
else:
    css_object_id = res_list[0]['css_object_id']
    print(css_object_id)
# To get incidents as well
test_incident = metadata_collection.find_one({'css_object_id': css_object_id})
backtrack_days = 60
projection = {
    "css_object_id": True,
    "prod_version": True,
    "customer_no": True,
    "customer_name": True,
    "component": True,
    "description": True,
    "incident_created_at_dt": True,
    "solns": True
}
query = {
    'prod_version': test_incident['prod_version'],
    'incident_created_at_dt': {
        '$gt': test_incident['incident_created_at_dt'] - timedelta(days=backtrack_days),
        '$lt': test_incident['incident_created_at_dt']
    },
    'customer_no': test_incident['customer_no'],
    'installation_no': test_incident['installation_no']
}
test_df = pd.DataFrame(metadata_collection.find(query, projection))
test_df['css_object_id']

002075129500004547572022


0     002075129400003203322022
1     002075129500003089632022
2     002075129500003099102022
3     002075129500003235062022
4     002075129400003324162022
5     002075129400003355412022
6     002075129400003396132022
7     002075129400003396902022
8     002075129500003262802022
9     002075129500003269362022
10    002075129500003370332022
11    002075129400003527732022
12    002075129500003549522022
13    002075129400003708932022
14    002075129400003828372022
15    002075129400003931292022
16    002075129500004290752022
17    002028376700036101062022
Name: css_object_id, dtype: object

In [178]:
infodocs_metadata_cloud = mongo_client['IBSO']['infodocs_metadata_cloud_only']
# Scenario 1
release_2202_df = pd.DataFrame(infodocs_metadata_cloud.find({'prod_version': 'SAP S/4HANA CLOUD 2202'}))
duplets_to_remove = release_2202_df[['customer_no', 'installation_no']].drop_duplicates()
all_cloud_df = pd.DataFrame(infodocs_metadata_cloud.find({}))
ind_to_remove = pd.Int64Index([])
for row in duplets_to_remove.iloc:
    ind_to_remove = ind_to_remove.join(all_cloud_df[(all_cloud_df['customer_no'] == row['customer_no']) & (all_cloud_df['installation_no'] == row['installation_no'])].index, how='outer')
ind_to_remove
scenario_1_df = all_cloud_df.drop(index=ind_to_remove)
# Scenario 2
latest_actual_dt = list(infodocs_metadata_cloud.find({'prod_version': 'SAP S/4HANA CLOUD 2202'}).sort('incident_created_at_dt', -1).limit(1))[0]['incident_created_at_dt']
latest_dt = latest_actual_dt - timedelta(days=61)
scenario_2_df = pd.DataFrame(infodocs_metadata_cloud.find({'incident_created_at_dt': {
    '$lt': latest_dt
}}))
# Scenario 3
scenario_3_df = pd.DataFrame(infodocs_metadata_cloud.find({}))








In [181]:
scenario_1_df.shape

(41532, 14)

In [183]:
scenario_1_df.head(30)

_id             component             css_object_id  \
0   635002305b3257a4b003b47d  XX-SER-SAPSMP-LAUNCH  002075129400004638912020   
1   635002305b3257a4b003b47e            LOD-MDM-DE  002075129500004950352019   
2   635002305b3257a4b003b47f       CEC-C4F-CPXF-L2  002075129500003794572019   
3   635002305b3257a4b003b480           CEC-SAL-LIT  002075129400002792242019   
4   635002305b3257a4b003b481       CEC-C4F-CPXF-L2  002075129400002588102019   
5   635002305b3257a4b003b482       CEC-C4F-CPXF-L2  002075129400002290692019   
6   635002305b3257a4b003b483       CEC-C4F-CPXF-L2  002075129500002206962019   
7   635002305b3257a4b003b484            BC-NEO-CPT  002075129500000918872019   
8   635002305b3257a4b003b485          CRM-S4-IC-BF  002075129500005570462018   
9   635002305b3257a4b003b486         MM-IM-GF-AUTH  002075129400007058512019   
10  635002305b3257a4b003b487        BC-SRV-APS-COM  002024211700000000182020   
11  635002305b3257a4b003b488            CA-GTF-MIG  002075129400001939352019   
17  635002305b3257a4b003b48e           CEC-MKT-CPG  002028376500001332962018   
18  635002305b3257a4b003b48f        XX-S4C-OPR-SRV  002075129500002445652018   
19  635002305b3257a4b003b490         LOD-HCI-PI-RT  002075129400002307912018   
20  635002305b3257a4b003b491      FIN-FSCM-CLM-BAM  002075129500002074342018   
21  635002305b3257a4b003b492           CEC-MKT-EXT  002075129400001986622018   
22  635002305b3257a4b003b493          SD-BIL-IV-PC  002075129400001844522018   
23  635002305b3257a4b003b494       CEC-MKT-CPG-EXE  002075129500001801212018   
24  635002305b3257a4b003b495         LOD-HCI-PI-RT  002075129500001799652018   
25  635002305b3257a4b003b496        XX-S4C-OPR-SRV  002075129400002448752018   
26  635002305b3257a4b003b497            CA-GTF-BUM  002075129500002311372018   
27  635002305b3257a4b003b498       XX-PROJ-CSC-HYB  002075129500002241512018   
28  635002305b3257a4b003b499           CEC-MKT-LPC  002075129500002157352018   
29  635002305b3257a4b003b49a         CEC-MKT-DM-CE  002024211700000012502018   
30  635002305b3257a4b003b49b        XX-S4C-OPR-SRV  002075129400001825692018   
31  635002305b3257a4b003b49c           CEC-MKT-MEM  002075129400001825522018   
32  635002305b3257a4b003b49d       CEC-MKT-INT-CSL  002075129400001721832018   
35  635002305b3257a4b003b4a0        XX-S4C-OPR-INC  002075129400002676852018   
36  635002305b3257a4b003b4a1         CEC-MKT-DM-IC  002075129400002357422018   

                          customer_name customer_no  \
0   SAP Test Account - aPaul Pharma Tes  0001208936   
1   SAP Test Account - aPaul Pharma Tes  0001208936   
2   SAP Test Account - aPaul Pharma Tes  0001208936   
3   SAP Test Account - aPaul Pharma Tes  0001208936   
4   SAP Test Account - aPaul Pharma Tes  0001208936   
5   SAP Test Account - aPaul Pharma Tes  0001208936   
6   SAP Test Account - aPaul Pharma Tes  0001208936   
7                      EOH Mthombo Ltd.  0001006511   
8                      EOH Mthombo Ltd.  0001006511   
9                     Pure Romance, LLC  0001644319   
10                Western Stone & Metal  0000753909   
11                     EOH Mthombo Ltd.  0001006511   
17                 UNIFAR, S.A. de C.V.  0001348433   
18  Hilti Aktiengesellschaft Hauptverwa  0000010039   
19  Hilti Aktiengesellschaft Hauptverwa  0000010039   
20  Hilti Aktiengesellschaft Hauptverwa  0000010039   
21  Hilti Aktiengesellschaft Hauptverwa  0000010039   
22  Hilti Aktiengesellschaft Hauptverwa  0000010039   
23  Hilti Aktiengesellschaft Hauptverwa  0000010039   
24  Hilti Aktiengesellschaft Hauptverwa  0000010039   
25                              BASF SE  0000010161   
26                              BASF SE  0000010161   
27                              BASF SE  0000010161   
28                              BASF SE  0000010161   
29                              BASF SE  0000010161   
30                              BASF SE  0000010161   
31                              BASF SE  0000010161   
32  